### DISCUSSION TITLE
tests/DCPS/MetaStruct/MetaStructTest failes on Yocto

### DISCUSSION THREAD

**Discussion by vermaete** [OpenDDS/OpenDDS#4620] [Q&A]
On a Yocto build with meta-opendds the test 'MetaStructTest' is failing.

```
root@rpi4-opendds:/usr/share/DDS_ROOT/tests/DCPS/MetaStruct# ./run_test.pl 
run_test_i<FinalStruct>(Unaligned CDR)
ERROR: Serialized type of a.w does not match. expected 1 received 2
run_test_i<FinalStruct>(XCDR2)
ERROR: Serialized type of a.w does not match. expected 1 received 2
run_test_i<AppenableStruct>(Unaligned CDR)
ERROR: Serialized type of a.w does not match. expected 1 received 2
run_test_i<AppenableStruct>(XCDR2)
ERROR: Serialized type of a.w does not match. expected 1 received 2
run_test_i<MutableStruct>(Unaligned CDR)
ERROR: Serialized type of a.w does not match. expected 1 received 2
run_test_i<MutableStruct>(XCDR2)
ERROR: Serialized type of a.w does not match. expected 1 received 2
ERROR: test returned 1
```

https://github.com/OpenDDS/OpenDDS/blob/master/tests/DCPS/MetaStruct/MetaStructTest.cpp#L128
https://github.com/OpenDDS/OpenDDS/blob/master/tests/DCPS/MetaStruct/MetaStructTest.cpp#L173

Version of OpenDDS: 3.28.0
Version of gcc: 13.2  !

The same test at Debian with GCC12.2 is passing.

I assume it's related to the size of wchar_t?

Br,

**Comment by mitza-oci** [comment]
What is the sizeof wchar_t on your platform?  Is ACE_WCHAR_T the same as wchar_t?

**Reply by mitza-oci** [reply]
The generated code is constructing a `DCPS::Value` (FilterEvalutator.h) object from the ACE_WCHAR_T.  There may be platform differences about which constructor is selected since wchar is not one of the cases -- it probably should be.

**Reply by vermaete** [reply]
Hi,

Does this answer your question?

```
root@rpi4-opendds:~# cat size.cpp 
#include <iostream>
#include <ace/Basic_Types.h>

int main() {
    std::cout << "Size of wchar_t: " << sizeof(wchar_t) << " bytes" << std::endl;
    std::cout << "Size of ACE_WCHAR_T: " << sizeof(ACE_WCHAR_T) << "bytes" << std::endl;
    return 0;
}
```

```
root@rpi4-opendds:~# ./a.out 
Size of wchar_t: 4 bytes
Size of ACE_WCHAR_T: 4bytes
```

**Reply by mitza-oci** [reply]
So that's the same as all Linux systems as far as I know.  The difference seems to be in how the compiler selects signed/unsigned integers to convert it into.  Or something else is going on.
